### ISSUE TITLE
[Feature] A better way to define Jobs environment variables

### ISSUE THREAD

**Issue by mohaidoss** [PrefectHQ/prefect-helm#415]
Currently, we can't pass Secrets or ConfigMap kube objects to the basejobtemplate.
This either requires to explicitly expose all values (secrets too) in a `basejobtemplate` config, or during the creation of a prefect deployment.

The **first option** is an anti-pattern as we don't benefit from the defined k8s secret/configmap objects.
As for the **second option**, it is not a developer responsibility to manage or know the value of env variables, nor the k8s object name.

This topic is somehow recurring on your slack:
- [Discussion 1](https://linen.prefect.io/t/23212158/hi-guys-still-new-to-prefect-3-and-i-was-wondering-if-there-)
- [Discussion 2](https://linen.prefect.io/t/23170850/ulva73b9p-in-prefect-3-in-the-kubernetes-worker-chart-what-d#c05e70f4-34e6-4146-89f2-88ab77c17aeb)
- [Discussion 3](https://linen.prefect.io/t/23170850/ulva73b9p-in-prefect-3-in-the-kubernetes-worker-chart-what-d#2dc8d732-7600-4228-afc8-272001580937)

**Comment by parkedwards** [comment]
hey @mohaidoss - thank you for raising this issue! 

> Currently, we can't pass Secrets or ConfigMap kube objects to the basejobtemplate.

You actually can - a base job template affords you both the higher-level abstraction of `variables` as well as a lower-level configuration of the Job manifest directly.  For example:

```
➜ prefect work-pool get-default-base-job-template --type kubernetes | grep job_manifest -A 29
```

^ searching for the `job_manifest` string in the generated base job template for the `--type kubernetes` (assuming this is the type of worker you're using because you're mentioning k8s objects

```
➜ prefect work-pool get-default-base-job-template --type kubernetes | grep job_manifest -A 29
    "job_manifest": {
      "apiVersion": "batch/v1",
      "kind": "Job",
      "metadata": {
        "labels": "{{ labels }}",
        "namespace": "{{ namespace }}",
        "generateName": "{{ name }}-"
      },
      "spec": {
        "backoffLimit": 0,
        "ttlSecondsAfterFinished": "{{ finished_job_ttl }}",
        "template": {
          "spec": {
            "parallelism": 1,
            "completions": 1,
            "restartPolicy": "Never",
            "serviceAccountName": "{{ service_account_name }}",
            "containers": [
              {
                "name": "prefect-job",
                "env": "{{ env }}",
                "image": "{{ image }}",
                "imagePullPolicy": "{{ image_pull_policy }}",
                "args": "{{ command }}"
              }
            ]
          }
        }
      }
    },
```

From here, you'll see that you have the ability to control container level settings, as you would on a k8s Job spec directly.  The default template uses the `env` template variable directly - the intention is to allow these environment variables to be set from a consumer level, like the Work Pool UI or a Deployment-level override.  For example, @zzstoatzz [from our team commented in the linen link you posted here](https://linen.prefect.io/t/23212158/hi-guys-still-new-to-prefect-3-and-i-was-wondering-if-there-); there's more imperative control to setting these env vars on a per-deployment level in prefect.yaml.  

If you feel that this isn't how you want to set these values, then you can absolutely set them in the base job template here

To drive this home, I've deployed a k8s worker with a custom base job template that looks like this - a custom env var that uses a `secretKeyRef`
```
// ... other stuff

    "job_manifest": {
      "apiVersion": "batch/v1",
      "kind": "Job",
      "metadata": {
        "labels": "{{ labels }}",
        "namespace": "{{ namespace }}",
        "generateName": "{{ name }}-"
      },
      "spec": {
        "backoffLimit": 0,
        "ttlSecondsAfterFinished": "{{ finished_job_ttl }}",
        "template": {
          "spec": {
            "parallelism": 1,
            "completions": 1,
            "restartPolicy": "Never",
            "serviceAccountName": "{{ service_account_name }}",
            "containers": [
              {
                "name": "prefect-job",
                "env": [
                  {
                    "name": "SOMETHING-ELSE",
                    "valueFrom": {
                      "secretKeyRef": {
                        "name": "prefect-api-key",
                        "key": "key"
                      }
                    }
                  }
                ],
                "image": "{{ image }}",
                "imagePullPolicy": "{{ image_pull_policy }}",
                "args": "{{ command }}"
              }
            ]
          }
        }
      }
// ...
```

and it'll populate the Job's pod:

```
➜ k describe pod platinum-capybara-qnqnm-mv47c | grep 'SOMETHING-ELSE'
      SOMETHING-ELSE:                    <set to the key 'key' in secret 'prefect-api-key'>  Optional: false
```

so you can see how you can easily set your own `valueFrom` with a secret or configmap.  If you were so inclined, you can also just add your own `envFrom` and mount entire configmaps as env vars

TL;DR:
- base job templates give you low level Job spec control, including adding your own `env` values
- You can use the default template's definition to pass in `env` variables from a work pool (UI) or deployment (prefect.yaml).  If you want to set it at the template level, you can customize your base job template and push it up
- You can extend your base job template to include `envFrom` if that's how you'd like to load your variables

Hope this helps!

**Comment by mohaidoss** [comment]
@parkedwards Thank you for bringing this to light, this is exactly what I was looking for.
However in no way this method of assigning variables were documented or recommended (other than using it in `prefect.yaml` or during prefect deployment creation).

I am interested in using `envFrom` what would the template look like in that case ? Is this a correct template ?

`job_template.json`
```json
// ... other stuff
{
  "containers": [
    {
      "name": "prefect-job",
      "envFrom": [
        {
          "configMapRef": {
            "name": "container-env-cm"
          }
        },
        {
          "secretRef": {
            "name": "prefect-api-key"
          }
        }
      ],
      "image": "{{ image }}",
      "imagePullPolicy": "{{ image_pull_policy }}",
      "args": "{{ command }}"
    }
  ]
}
// ...
```

**Comment by parkedwards** [comment]
@mohaidoss exactly. do give this a try and please report back if you run into issues
